### PR TITLE
Added multivariate normal distribution

### PR DESCRIPTION
**Pull request by sidravi1**
## Overview

Add MvNormal distribution. Uses `jax.scipy.multivariate_normal` for `logpdf` and `jax.random.multivariate_normal` for sampling.

## Quick tests

```
>>> import jax
>>> import jax.numpy as np
>>> import mcx
>>>
>>> rng_key = jax.random.PRNGKey(0)
>>> mvn = mcx.distributions.MvNormal(np.array([1, 2]), np.array([[1.0, 0.2], [0.2, 1.0]]))
>>> mvn.sample(rng_key)
DeviceArray([0.21523398, 2.6821878 ], dtype=float32)
>>> mvn.sample(rng_key, (3,))
DeviceArray([[1.1878438 , 0.78015494],
             [1.649418  , 3.3537068 ],
             [1.2444699 , 1.9338173 ]], dtype=float32)
>>> mvn.logpdf(np.array([1, 2]))
DeviceArray(-1.817466, dtype=float32)
>>> mvn.logpdf(np.array([[1, 2], [2, 1]]))
DeviceArray([-1.817466, -3.067466], dtype=float32)
```

Not sure the batching is broadcasting correctly though:

```
>>> mvn2 = mcx.distributions.MvNormal(
...     np.array([[1, 2], [3, 2]]), np.array([[1.0, 0.2], [0.2, 1.0]])
... )
>>> mvn2.sample(rng_key)
DeviceArray([[2.8160858, 1.6235838],
             [3.339889 , 1.5439482]], dtype=float32)
>>> mvn2.sample(rng_key, (3, 2))
DeviceArray([[[[ 1.3675394 ,  1.183653  ],
               [ 0.99355793,  1.7560301 ]],

              [[ 1.1323344 ,  0.7474071 ],
               [ 2.5944324 ,  0.161587  ]]],


             [[[-0.35665548,  2.5218964 ],
               [ 2.6202202 ,  2.0067666 ]],

              [[-0.89568627,  1.4151701 ],
               [ 3.20252   ,  3.3841357 ]]],


             [[[ 0.3996758 ,  0.86409783],
               [ 4.5410695 ,  2.3596075 ]],

              [[ 1.0302643 ,  3.2970448 ],
               [ 3.6156623 ,  3.787726  ]]]], dtype=float32)
>>> mvn.logpdf(np.array([1, 2]))
DeviceArray(-1.817466, dtype=float32)
>>> mvn.logpdf(np.array([[1, 2], [2, 1]]))
DeviceArray([-1.817466, -3.067466], dtype=float32)
```

## Pending

Proper test cases
Fixing shape issues

### CODE REVIEW

File: mcx/distributions/mvnormal.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+class MvNormal(Distribution):
+    params_constraints = {
+        "mu": constraints.real_vector,
+        "covariance_matrix": constraints.positive_definite,
+    }
+    support = constraints.real_vector
+
+    def __init__(self, mu, covariance_matrix):
+        self.event_shape = broadcast_batch_shape(np.shape(mu))
```
Comment (rlouf) [inline mcx/distributions/mvnormal.py:17]: Shapes are trickier with multivariate distributions. `event_shape` refers to the shape of one sample from the distribution; it is here different from 1. It is indeed specified by `mu`, which is necessarily a vector. But we also want to determine the `batch_shape` (number of iid samples returned) with `mu`.

So say we want to sample from 4 iid random variables with a Multivariate Normal distribution of mean vector `[1., 1.]`. Naturally we want to write something like:

`mu = [[1,1], [1,1], [1,1], [1,1]]`

of shape `(4, 2)`. We can thus choose the *convention* `event_shape = broadcast_batch_shape(np.shape(mu)[-1])`.

File: mcx/distributions/mvnormal.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        self.batch_shape = broadcast_batch_shape(
```
Comment (rlouf) [inline mcx/distributions/mvnormal.py:18]: And as you implied, we may also want to define iid random variables with the same mean vector but different covariance matrices; `batch_shape` thus also depends on the shape of `covariance_matrix`. But we want to ignore the last dimension of the broadcasted shape since it will correspond to the `event_shape`. So:

`batch_shape = broadcast_batch_shape(np.shape(mu), np.shape(covariance_matrix))[:-2]`

File: mcx/distributions/mvnormal.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    def sample(self, rng_key, sample_shape=()):
+        shape = sample_shape + self.batch_shape + self.event_shape
+        draws = random.multivariate_normal(
+            rng_key, mean=self.mu, cov=self.covariance_matrix, shape=shape
+        )
+        return draws
+
+    # no need to check on support ]-infty, +infty[
+    def logpdf(self, x):
+        if (x[0].shape != self.mu.shape) and (x.shape != self.mu.shape):
```
Comment (rlouf) [inline mcx/distributions/mvnormal.py:34]: No need to raise an exception if `jax.scipy` does!

File: mcx/distributions/mvnormal.py (hunk 2 above)
Comment (rlouf) [reply mcx/distributions/mvnormal.py:18]: Note that we should also check that the shapes of `mu` and `covariance_matrix` are consistent i.e. that `np.shape(covariance_matrix)[-2:-1]` is equal to `event_shape`. And raise an exception if that is not the case.

File: mcx/distributions/constraints.py [hunk 4]
```diff
@@ -175,14 +177,65 @@ def __call__(self, x):
...
+
+    def __call__(self, x):
+        jnp = numpy if isinstance(x, (numpy.ndarray, numpy.generic)) else np
+        symmetric = jnp.all(jnp.all(x == jnp.swapaxes(x, -2, -1), axis=-1), axis=-1)
+        # check for the smallest eigenvalue is positive
+        positive = jnp.linalg.eigh(x)[0][..., 0] > 0
+        return symmetric & positive
+
+
+class _LowerCholesky(Constraint):
```
Comment (rlouf) [inline mcx/distributions/constraints.py:216]: Don't forget to remove `LowerCholesky` at the end if we choose not to implement it.

**Review by rlouf — COMMENTED** [review]
That's a great start! I left some comments regarding the shapes in the code. It took me a while to get used to the shape system, and that was with a constant event shape of 1 :) If you look at the existing tests for distributions, what I did was write down the shapes I expected in different situations and added tests for that; modified the code until tests passed.

File: mcx/distributions/mvnormal.py (hunk 1 above)
Comment (sidravi1) [reply mcx/distributions/mvnormal.py]: That was helpful. Thanks! 
That + Eric Ma's blogpost + some mucking around and i *think* i have it.

**Comment by sidravi1** [comment]
I think I'm mostly there with shape. Will run a bunch of tests tomorrow and if it's all kosher, I'll put it up for another review.

**Comment by sidravi1** [comment]
Added some test cases. I think shape stuff is looking ok now.

Batching when using jax.scipy's multivariate_normal's logp gives `inf`s. Not sure if I'm doing something stupid. I raised [this](https://github.com/google/jax/issues/5570).

File: mcx/distributions/constraints.py [hunk 5]
```diff
@@ -175,14 +177,52 @@ def __call__(self, x):
...
+            )
+        result = result.all(-1)
+        return result
+
+
+class _PositiveDefinite(Constraint):
+    event_dim = 2
+
+    def __call__(self, x):
+        jnp = numpy if isinstance(x, (numpy.ndarray, numpy.generic)) else np
```
Comment (rlouf) [inline mcx/distributions/constraints.py:209]: Does it otherwise fail if you pass a numpy array?

File: mcx/distributions/constraints.py [hunk 6]
```diff
@@ -175,14 +177,52 @@ def __call__(self, x):
...
+            return result
+        elif np.ndim(result) < self.ndims:
+            raise ValueError(
+                (f"Expected value.dim() >= {self.ndims}" " but got {np.ndims{result}}")
+            )
+        result = result.all(-1)
+        return result
+
+
+class _PositiveDefinite(Constraint):
```
Comment (rlouf) [inline mcx/distributions/constraints.py:205]: nitpicking: can you add a __str__ method?

File: mcx/distributions/constraints.py (hunk 5 above)
Comment (rlouf) [reply mcx/distributions/constraints.py:209]: This `jnp` / `np` mess reminds me that I need to switch to `import jax.numpy as jnp` everywhere :)

File: mcx/distributions/mvnormal.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+class MvNormal(Distribution):
+    params_constraints = {
+        "mu": constraints.real_vector,
+        "covariance_matrix": constraints.positive_definite,
+    }
+    support = constraints.real_vector
+
+    def __init__(self, mu, covariance_matrix):
+
+        (mu_event_shape,) = np.shape(mu)[-1:]
```
Comment (rlouf) [inline mcx/distributions/mvnormal.py:18]: I like the explicit names!

File: mcx/distributions/mvnormal.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        covariance_event_shape = np.shape(covariance_matrix)[-2:]
+        if (mu_event_shape, mu_event_shape) != covariance_event_shape:
+            raise ValueError(
+                (
+                    f"event shape {mu_event_shape} of mu does not match event "
+                    f"shape {covariance_event_shape} of covariance_matrix"
```
Comment (rlouf) [inline mcx/distributions/mvnormal.py:24]: I suspect that > 75% of users will run into shape errors with `MvNormal`. Maybe we should phrase it in terms of something they know (dimensions) instead of event shape (only devs do and should know about this):

"the number of dimensions {mu_event_shape} implied by the mean vector `mu` you provided "
"does not match the number of dimensions implied by the shape {covariance_event_shape} "
"of the covariance matrix you provided. "

What do you think?

File: tests/distributions/mvn_test.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+from mcx.distributions import MvNormal
+
+
+@pytest.fixture
+def rng_key():
+    return random.PRNGKey(0)
+
+
+#
+# LOGPDF SHAPES
```
Comment (rlouf) [inline tests/distributions/mvn_test.py:14]: kudos for testing the logpdf shapes more thoroughly than I did!

**Review by rlouf — COMMENTED** [review]
Great job! I am particularly impressed by the thorough testing. I only have minor comments.

It looks like tests are partly failing because the versions of JAX and JAXlib are not compatible. I'll merge a patch right away, could you rebase your branch on master afterwards?

Also, are the tests affected by the bug on JAX's side?

**Comment by rlouf** [comment]
patched!

File: mcx/distributions/constraints.py (hunk 5 above)
Comment (sidravi1) [reply mcx/distributions/constraints.py:209]: I am basically ripping off the code from [`numpyro`](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/constraints.py). This is how they define constraints there. Should I test what happens without this line? Actually, should we have test cases for constraints?

File: mcx/distributions/constraints.py (hunk 6 above)
Comment (sidravi1) [reply mcx/distributions/constraints.py:205]: Ah yes! For sure

File: mcx/distributions/constraints.py (hunk 5 above)
Comment (sidravi1) [reply mcx/distributions/constraints.py:209]: Also, want me to create an issue for the `jnp` business?

File: mcx/distributions/mvnormal.py (hunk 8 above)
Comment (sidravi1) [reply mcx/distributions/mvnormal.py:24]: I like this. Ok will fix.

**Comment by sidravi1** [comment]
Made the fixes!

Re: the jax bug. The tests are not affected since they are only checking shape. The jax bug leads to `inf` being returned but the shape is correct.

File: mcx/distributions/constraints.py (hunk 5 above)
Comment (rlouf) [reply mcx/distributions/constraints.py:212]: Yes if you could check what happens without this line it would be great. No need to add a test for constraints here, just making sure it works once should be enough!

And yes opening an issue marking it as "good first issue" would be great 😊

**Comment by rlouf** [comment]
> Re: the jax bug. The tests are not affected since they are only checking shape. The jax bug leads to `inf` being returned but the shape is correct.

Ok, hopefully it will be fixed before the first release.

> Btw... Do you want me to squash all my merges so the commit history doesn't get messy?

Yes please!

File: mcx/distributions/constraints.py (hunk 5 above)
Comment (sidravi1) [reply mcx/distributions/constraints.py:212]: Your intuition was right. Works without that line and just using jax.numpy for all.
Will clean it up.

Comment (sidravi1) [reply mcx/distributions/constraints.py:212]: I created issue #72 but don't have permissions to add labels etc. You might have to do that.

**Comment by rlouf** [comment]
Looks good to me, merging. Great job!

**Comment by rlouf** [comment]
Addressed one item in #65
